### PR TITLE
[Fixed JENKINS-14591] migrate parameter options to extensible behavior

### PR DESCRIPTION
**Pull request by dalvizu**
[JENKINS-14591](https://issues.jenkins-ci.org/browse/JENKINS-14591)

Rename config to 'row header' and 'column header' options. Provide different
options based on existing behavior and requests from users:
- No column header
- Simple headers
- Headers with parameters
- Header with all build variables

Refactored these options to encapsulate the UI logic better - no more large if / else blocks in `bpp.jelly`, they belong to their own classes now. Users can now provide their own `@Extension` of `PipelineHeaderExtension` to customize, following open-closed principle.

### CODE REVIEW

**Comment by dalvizu** [comment]
cc @recena

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/configure-entries.jelly [hunk 1]
```diff
@@ -6,42 +6,56 @@
...
+        <f:entry title="Console Output Link Style" field="consoleOutputLinkStyle">
+            <f:select />
+        </f:entry>
+    </f:section>
+
+    <script type="text/javascript">
+        // dropdownDescriptorSelectors will fill items up but don't update the layout, so need
+        // to do that explicitly
+        setTimeout(function(){
+            layoutUpdateCallback.call();
```
Comment (dalvizu) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/configure-entries.jelly]: This was annoying AF - without this, the 'save / ok' button will overlap on the fields. If there's a better way to do it I'm open to it.

**Comment by recena** [comment]
@dalvizu We need to merge this https://github.com/jenkinsci/build-pipeline-plugin/pull/95 before.

File: pom.xml [hunk 2]
```diff
@@ -241,6 +241,7 @@
                         <goals>
                             <goal>check</goal>
                         </goals>
+                        <phase>process-classes</phase>
```
Comment (recena) [inline pom.xml]: unrelated change

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java [hunk 3]
```diff
@@ -692,6 +814,21 @@ public ListBoxModel doFillConsoleOutputLinkStyleItems() {
...
+     */
+    public interface Function<F, T> {
+        /**
+         *
+         * @param input
+         *  an input
+         * @return
+         *  a result
+         */
+        T apply(F input);
```
Comment (recena) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java]: This javadoc don't make sense.

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * .jelly files to inherit for children. See resource files:
+ *
+ * <ul>
+ *   <li><tt>src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader/projectCardTemplate.jelly</tt></li>
+ *   <li><tt>src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader/rowHeader.jelly</tt></li>
+ * </ul>
+ *
+ * @author dalvizu
+ */
+public abstract class AbstractNameValueHeader
```
Comment (recena) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader.java]: Why not one-line?

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeader.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Pull values from build variables, hiding sensitive ones
+ *
+ * @author dalvizu
+ */
+@Extension
+public class BuildVariablesHeader
```
Comment (recena) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeader.java]: ditto.

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/ParametersActionHeader.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * Provide parameters only from the ParametersAction and nothing else
+ *
+ * See issue JENKINS-14591
+ *
+ * @see <a href="https://github.com/jenkinsci/build-pipeline-plugin/pull/74">Pull request #74</a>
+ * @author dalvizu
+ */
+@Extension(optional = true)
+public class ParametersActionHeader
```
Comment (recena) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/ParametersActionHeader.java]: DItto

File: src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeaderTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ * @author dalvizu
+ */
+public class BuildVariablesHeaderTest {
+
+    private Map<String, String> buildVariableMap;
+
+    @Before
+    public void setup()
+    {
+        buildVariableMap = new HashMap<String, String>();
```
Comment (recena) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeaderTest.java]: `public void setup() {`

File: pom.xml (hunk 2 above)
Comment (dalvizu) [inline pom.xml]: Done

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader.java (hunk 4 above)
Comment (dalvizu) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader.java]: Done

Comment (dalvizu) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/AbstractNameValueHeader.java]: Done

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeader.java (hunk 5 above)
Comment (dalvizu) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeader.java]: Done

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/ParametersActionHeader.java (hunk 6 above)
Comment (dalvizu) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/ParametersActionHeader.java]: Done

File: src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeaderTest.java (hunk 7 above)
Comment (dalvizu) [inline src/test/java/au/com/centrumsystems/hudson/plugin/buildpipeline/extension/BuildVariablesHeaderTest.java]: Done - surprised checkstyle didn't find this. It was yelling at me a lot

File: pom.xml (hunk 2 above)
Comment (dalvizu) [inline pom.xml]: Done - I think it is worth doing (checkstyle after tests is a bit of a bummer), but will put that in another PR

**Comment by dalvizu** [comment]
@recena are you ready to merge #95?

File: src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java (hunk 3 above)
Comment (dalvizu) [inline src/main/java/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView.java]: Updated - this is a JDK8 function, but we're not requiring our users to have JDK8, so I copied it

**Comment by dalvizu** [comment]
Finally got around to fixing the merge conflict! Testing looks good. @recena you've already reviewed, but it's so long ago that I want to check: are you 👍 for me to merge?
